File: rust/crates/spore-core/examples/e2e_agent.rs
//! End-to-end CLI agent harness + scenario suite (issue #57).
//!
//! One shared example binary that drives the *complete* harness loop against a
//! real local model (Ollama) through a `HarnessBuilder`-assembled harness with
//! real tools (read/write/list/bash + a deliberately-failing tool), the
//! `StandardCompactionAdapter`, and the durable-outbox observability provider.
//! The scenario is selected by a CLI arg.
//!
//! ## Scenarios
//!
//! - `s1` — multi-step / multi-tool: read input.txt → uppercase → write
//!   output.txt → read back + confirm.
//! - `s2` — multi-turn: run twice with the same `SessionId`, carrying session
//!   state across turns.
//! - `s3` — live compaction: a seeded small window + long history fires the
//!   compaction adapter mid-run; the token-accounting fix lets it compact,
//!   continue, and compact again.
//! - `s4` — tool failure + recovery: call `flaky_op` (recoverable error), then
//!   write recovered.txt explaining the adaptation.
//!
//! ## Run recipe (live, against a local model + observability stack)
//!
//! ```sh
//! # 1. Start Ollama and pull a tool-capable model.
//! ollama serve &              # or run the Ollama app
//! ollama pull llama3.2        # default model; passes the #41 capability guard
//!
//! # 2. (optional) Start the local observability stack and forward traces.
//! #    See observability/ for the compose stack (Tempo + Loki + Grafana).
//! export SPORE_OTLP_ENDPOINT=http://localhost:4317
//!
//! # 3. Run a scenario. Prompt/model/endpoint/workspace come from args+env.
//! cargo run -p spore-core --example e2e_agent -- s1 --model llama3.2
//! cargo run -p spore-core --example e2e_agent -- s2
//! cargo run -p spore-core --example e2e_agent -- s3
//! cargo run -p spore-core --example e2e_agent -- s4
//!
//! # 4. Verify the grouped trace in Tempo (the run prints the trace_id):
//! curl -s http://localhost:3200/api/traces/<trace_id> | jq '.batches | length'
//! #    For S3, spot-check a Compaction span appears mid-trace.
//! ```
//!
//! Environment variables (all optional):
//! - `SPORE_OLLAMA_MODEL`     — default model id (overridden by `--model`).
//! - `SPORE_OLLAMA_BASE_URL`  — Ollama base url (default http://localhost:11434).
//! - `SPORE_OTLP_ENDPOINT`    — when set, forward spans to Tempo (issue #50).
//! - `SPORE_E2E_WORKSPACE`    — workspace root (default: a temp dir per run).
//!
//! ## Offline / hermetic mode
//!
//! `--mock` runs the same scenario builders against a scripted `MockAgent`,
//! requiring no Ollama or network. Mock support is gated behind the
//! `test-utils` feature:
//!
//! ```sh
//! cargo run -p spore-core --features test-utils --example e2e_agent -- s1 --mock
//! ```
//!
//! The hermetic CI assertions live in `tests/e2e_scenarios.rs`, which drive the
//! same `build_scenario` path with mock components.

use std::sync::Arc;

use spore_core::agent::{AgentId, ModelAgent};
use spore_core::cache_provider::OllamaCacheProvider;
use spore_core::context::CompactionConfig;
use spore_core::observability_outbox::{OutboxConfig, OutboxObservabilityProvider};
use spore_core::scenarios::{
    build_real_tool_registry, build_rich_context_manager, build_scenario, seed_compaction_state,
    CompleteOnFinalResponse, RealToolRegistry, ScenarioId,
};
use spore_core::{
    Agent, FullObservabilityProvider, Harness, HarnessContextManager, HarnessRunOptions,
    HarnessToolRegistry, LoopStrategy, OllamaModelInterface, RunResult, SandboxProvider, SessionId,
    SessionState, Task, TerminationPolicy, Timestamp, WorkspaceConfig, WorkspaceScopedSandbox,
};

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().skip(1).collect();
    let scenario = args
        .first()
        .and_then(|s| ScenarioId::parse(s))
        .unwrap_or_else(|| {
            eprintln!("usage: e2e_agent <s1|s2|s3|s4> [--model <id>] [--mock]");
            std::process::exit(2);
        });
    let mock = args.iter().any(|a| a == "--mock");
    let model_id = arg_value(&args, "--model")
        .or_else(|| std::env::var("SPORE_OLLAMA_MODEL").ok())
        .unwrap_or_else(|| "llama3.2".to_string());

    // Per-run session id so repeated runs don't collide in Tempo.
    let session_id = SessionId::new(format!(
        "e2e-{:?}-{}",
        scenario,
        Timestamp::now().as_str().replace([':', '.'], "-")
    ));

    // Workspace: a real directory the file tools operate inside.
    let workspace = std::env::var("SPORE_E2E_WORKSPACE")
        .map(std::path::PathBuf::from)
        .unwrap_or_else(|_| {
            std::env::temp_dir().join(format!("spore-e2e-{}", session_id.as_str()))
        });
    std::fs::create_dir_all(&workspace).expect("create workspace");
    prepare_workspace(scenario, &workspace);
    println!("workspace  : {}", workspace.display());

    // Observability root (honors SPORE_OTLP_ENDPOINT like observability_smoke).
    let obs_root = std::path::Path::new(env!("CARGO_MANIFEST_DIR")).join("../../../.spore");
    let endpoint = std::env::var("SPORE_OTLP_ENDPOINT").unwrap_or_default();
    if endpoint.trim().is_empty() {
        eprintln!("note: SPORE_OTLP_ENDPOINT is unset — writing JSONL only (no Tempo forwarding).");
    }

    let result = if mock {
        run_mock(scenario, &session_id).await
    } else {
        run_live(
            scenario,
            &session_id,
            &model_id,
            &workspace,
            obs_root.clone(),
        )
        .await
    };

    match result {
        RunResult::Success { turns, output, .. } => {
            println!("result     : Success ({turns} turns)");
            println!("output     : {output:?}");
        }
        other => {
            eprintln!("result     : {other:?}");
            std::process::exit(1);
        }
    }

    // Trace info: derive the on-disk JSONL path the outbox provider wrote and
    // surface the trace_id so the user can jump straight to Grafana/Tempo. Mock
    // runs use a no-op observability provider and write no JSONL, so guard the
    // read and print a note instead of panicking. S2 runs twice under one
    // session_id (one trace dir) — printing the single id once is sufficient.
    let sid = session_id.as_str();
    println!("session_id : {sid}");
    let trace_path = obs_root.join(format!("sessions/{sid}/trace.jsonl"));
    if mock {
        println!("trace_path : (mock run — no trace)");
    } else {
        match std::fs::read_to_string(&trace_path) {
            Ok(body) => {
                let trace_id = body
                    .lines()
                    .next()
                    .and_then(|l| serde_json::from_str::<serde_json::Value>(l).ok())
                    .and_then(|v| v["trace_id"].as_str().map(str::to_string));
                println!("trace_path : {}", trace_path.display());
                match trace_id {
                    Some(id) => {
                        println!("trace_id   : {id}");
                        if !endpoint.trim().is_empty() {
                            println!(
                                "verify     : curl -s http://localhost:3200/api/traces/{id} | jq '.batches | length'"
                            );
                        }
                    }
                    None => println!("trace_id   : (trace file empty or unparseable)"),
                }
            }
            Err(_) => {
                println!("trace_path : {} (not found)", trace_path.display());
            }
        }
    }
}

/// Run the scenario against a live Ollama model.
async fn run_live(
    scenario: ScenarioId,
    session_id: &SessionId,
    model_id: &str,
    workspace: &std::path::Path,
    obs_root: std::path::PathBuf,
) -> RunResult {
    if !OllamaModelInterface::supports_tools(model_id) {
        eprintln!(
            "warning: model `{model_id}` is not in the known tool-capable whitelist; \
             the harness will guard against tool use if the model lacks the capability."
        );
    }
    let base_url = std::env::var("SPORE_OLLAMA_BASE_URL")
        .unwrap_or_else(|_| OllamaModelInterface::DEFAULT_BASE_URL.to_string());
    let model = Arc::new(OllamaModelInterface::with_base_url(model_id, base_url));
    let agent: Arc<dyn Agent> = Arc::new(ModelAgent::new(AgentId::new("e2e-agent"), model.clone()));

    // Real tools + sandbox scoped to the workspace.
    let registry = build_real_tool_registry();
    let sandbox: Arc<dyn SandboxProvider> = Arc::new(
        WorkspaceScopedSandbox::new(WorkspaceConfig {
            root: workspace.to_path_buf(),
            allowed_paths: vec![],
            denied_paths: vec![],
            allowed_extensions: None,
            denied_extensions: vec![],
            read_only: false,
            max_file_size: 0,
        })
        .expect("build sandbox"),
    );
    let bridge = RealToolRegistry::new(registry, sandbox.clone());
    let tool_schemas = bridge.model_schemas();
    let tools: Arc<dyn HarnessToolRegistry> = Arc::new(bridge);

    // Compaction-capable context manager (small window for S3).
    let window_limit = if scenario == ScenarioId::S3 {
        200
    } else {
        128_000
    };
    let cfg = CompactionConfig {
        threshold: 0.80,
        preserve_recent_n: 2,
        head_tail_tokens: 64,
        offload_path: workspace.join(".spore/offload"),
        max_compaction_attempts: 2,
    };
    let context_manager: Arc<dyn HarnessContextManager> =
        build_rich_context_manager(model, Arc::new(OllamaCacheProvider), cfg);

    let termination: Arc<dyn TerminationPolicy> = Arc::new(CompleteOnFinalResponse);

    let obs: Arc<dyn FullObservabilityProvider> = Arc::new(OutboxObservabilityProvider::new(
        OutboxConfig::new(obs_root),
    ));

    let harness = build_scenario(
        scenario,
        agent,
        tools,
        sandbox,
        context_manager,
        termination,
        tool_schemas,
        Some(obs),
    );

    run_scenario(scenario, &harness, session_id, window_limit).await
}

/// Drive the scenario, including the S2 multi-turn and S3 seed.
async fn run_scenario(
    scenario: ScenarioId,
    harness: &impl spore_core::Harness,
    session_id: &SessionId,
    window_limit: u32,
) -> RunResult {
    let strategy = LoopStrategy::ReAct { max_iterations: 8 };

    match scenario {
        ScenarioId::S2 => {
            // Multi-turn: run twice with the same SessionId, carrying state.
            let task1 = Task::new(scenario.prompt(), session_id.clone(), strategy.clone());
            let r1 = harness.run(HarnessRunOptions::new(task1)).await;
            let state = match &r1 {
                RunResult::Success { .. } => SessionState::default(),
                other => {
                    eprintln!("S2 turn 1 did not succeed: {other:?}");
                    return r1;
                }
            };
            let task2 = Task::new(
                "Add a second TODO item to notes.md that references the first item you wrote. \
                 Use write_file with append=true. Reply DONE when finished.",
                session_id.clone(),
                strategy,
            );
            harness
                .run(HarnessRunOptions::new(task2).with_session_state(state))
                .await
        }
        ScenarioId::S3 => {
            let task = Task::new(scenario.prompt(), session_id.clone(), strategy);
            let mut state = SessionState::default();
            seed_compaction_state(
                &mut state,
                "deploy the payment service",
                session_id.clone(),
                task.id.clone(),
                window_limit,
                (window_limit as f32 * 0.82) as u32,
                12,
            );
            harness
                .run(HarnessRunOptions::new(task).with_session_state(state))
                .await
        }
        _ => {
            let task = Task::new(scenario.prompt(), session_id.clone(), strategy);
            harness.run(HarnessRunOptions::new(task)).await
        }
    }
}

/// Seed scenario-specific workspace files.
fn prepare_workspace(scenario: ScenarioId, workspace: &std::path::Path) {
    if scenario == ScenarioId::S1 {
        std::fs::write(
            workspace.join("input.txt"),
            "hello from the spore harness end to end scenario\n",
        )
        .expect("seed input.txt");
    }
}

fn arg_value(args: &[String], flag: &str) -> Option<String> {
    args.iter()
        .position(|a| a == flag)
        .and_then(|i| args.get(i + 1).cloned())
}

// ---------------------------------------------------------------------------
// Offline / mock mode (gated behind the test-utils feature).
// ---------------------------------------------------------------------------

#[cfg(feature = "test-utils")]
async fn run_mock(scenario: ScenarioId, session_id: &SessionId) -> RunResult {
    use spore_core::agent::mock::MockAgent;
    use spore_core::harness::testing::{
        AllowAllSandbox, AlwaysContinuePolicy, NoopContextManager, ScriptedToolRegistry,
    };
    use spore_core::{TokenUsage, ToolCall, ToolOutput, TurnResult};

    let _ = scenario;
    let agent = Arc::new(MockAgent::new(AgentId::new("mock")));
    let usage = TokenUsage {
        input_tokens: 10,
        output_tokens: 5,
        cache_read_tokens: None,
        cache_write_tokens: None,
    };
    agent.push(TurnResult::ToolCallRequested {
        calls: vec![ToolCall {
            id: "c1".into(),
            name: "read_file".into(),
            input: serde_json::json!({"path": "input.txt"}),
        }],
        usage,
    });
    agent.push(TurnResult::FinalResponse {
        content: "DONE".into(),
        usage,
    });
    let tools = Arc::new(ScriptedToolRegistry::new());
    tools.push(ToolOutput::Success {
        content: "contents".into(),
        truncated: false,
    });

    let harness = build_scenario(
        scenario,
        agent,
        tools,
        Arc::new(AllowAllSandbox),
        Arc::new(NoopContextManager),
        Arc::new(AlwaysContinuePolicy),
        vec![],
        None,
    );
    let task = Task::new(
        scenario.prompt(),
        session_id.clone(),
        LoopStrategy::ReAct { max_iterations: 5 },
    );
    harness.run(HarnessRunOptions::new(task)).await
}

#[cfg(not(feature = "test-utils"))]
async fn run_mock(_scenario: ScenarioId, _session_id: &SessionId) -> RunResult {
    eprintln!(
        "error: --mock requires the test-utils feature. Re-run with:\n  \
         cargo run -p spore-core --features test-utils --example e2e_agent -- <scenario> --mock"
    );
    std::process::exit(2);
}

File: typescript/packages/tools/src/bin/e2e-agent.ts
/**
 * End-to-end CLI agent harness + scenario suite (issue #57).
 *
 * One shared runnable entry that drives the *complete* harness loop against a
 * real local model (Ollama) through a `HarnessBuilder`-assembled harness with
 * real tools (read/write/list/bash + a deliberately-failing tool), the
 * `StandardCompactionAdapter`, and the durable-outbox observability provider.
 * The scenario is selected by a CLI arg.
 *
 * ## Scenarios
 *
 * - `s1` — multi-step / multi-tool: read input.txt → uppercase → write
 *   output.txt → read back + confirm.
 * - `s2` — multi-turn: run twice with the same `SessionId`, carrying session
 *   state across turns.
 * - `s3` — live compaction: a seeded small window + long history fires the
 *   compaction adapter mid-run; the token-accounting fix lets it compact,
 *   continue, and compact again.
 * - `s4` — tool failure + recovery: call `flaky_op` (recoverable error), then
 *   write recovered.txt explaining the adaptation.
 *
 * ## Run recipe (live, against a local model + observability stack)
 *
 * ```sh
 * # 1. Start Ollama and pull a tool-capable model.
 * ollama serve &              # or run the Ollama app
 * ollama pull llama3.2        # default model; passes the #41 capability guard
 *
 * # 2. (optional) Start the local observability stack and forward traces.
 * export SPORE_OTLP_ENDPOINT=http://localhost:4317
 *
 * # 3. Run a scenario. Prompt/model/endpoint/workspace come from args+env.
 * pnpm --filter @spore/tools e2e s1 --model llama3.2
 * pnpm --filter @spore/tools e2e s2
 * pnpm --filter @spore/tools e2e s3
 * pnpm --filter @spore/tools e2e s4
 *
 * # 4. Verify the grouped trace in Tempo (the run prints the trace_id):
 * curl -s http://localhost:3200/api/traces/<trace_id> | jq '.batches | length'
 * #    For S3, spot-check a Compaction span appears mid-trace.
 * ```
 *
 * Environment variables (all optional):
 * - `SPORE_OLLAMA_MODEL`     — default model id (overridden by `--model`).
 * - `SPORE_OLLAMA_BASE_URL`  — Ollama base url (default http://localhost:11434).
 * - `SPORE_OTLP_ENDPOINT`    — when set, forward spans to Tempo (issue #50).
 * - `SPORE_E2E_WORKSPACE`    — workspace root (default: a temp dir per run).
 *
 * ## Offline / hermetic mode
 *
 * `--mock` runs the same `buildScenario` builders against a scripted mock
 * agent, requiring no Ollama or network. The hermetic CI assertions live in
 * `packages/tools/tests/e2e-scenarios.test.ts`, which drive the same path.
 */

import { mkdirSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { tmpdir } from "node:os";

import {
  AgentId,
  type Agent,
  cacheProvider,
  context as coreContext,
  type HarnessRunOptions,
  type LoopStrategy,
  ModelAgent,
  newTask,
  observability as coreObs,
  OllamaModelInterface,
  OLLAMA_DEFAULT_BASE_URL,
  type RunResult,
  type SandboxProvider,
  SessionId,
  type SessionState,
  TaskId,
  type TerminationPolicy,
  type ToolSchema,
  type ToolRegistry as HarnessToolRegistry,
  type ContextManager as HarnessContextManager,
  emptySessionState,
  WorkspaceScopedSandbox,
  storage as coreStorage,
} from "@spore/core";

import {
  buildRealToolRegistry,
  buildRichContextManager,
  buildScenario,
  CompleteOnFinalResponse,
  parseScenarioId,
  RealToolRegistry,
  type ScenarioId,
  scenarioPrompt,
  seedCompactionState,
} from "../scenarios.js";

const { OutboxObservabilityProvider, outboxConfig } = coreObs;
const { OllamaCacheProvider } = cacheProvider;

function argValue(args: string[], flag: string): string | undefined {
  const i = args.indexOf(flag);
  return i >= 0 ? args[i + 1] : undefined;
}

function reactStrategy(maxIterations: number): LoopStrategy {
  return { kind: "re_act", max_iterations: maxIterations };
}

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const scenario = args[0] ? parseScenarioId(args[0]) : undefined;
  if (!scenario) {
    process.stderr.write(
      "usage: e2e-agent <s1|s2|s3|s4|s5> [--model <id>] [--mock]\n",
    );
    process.exit(2);
  }
  const mock = args.includes("--mock");
  const modelId =
    argValue(args, "--model") ?? process.env.SPORE_OLLAMA_MODEL ?? "llama3.2";

  const stamp = new Date().toISOString().replace(/[:.]/g, "-");
  const sessionId = SessionId.of(`e2e-${scenario}-${stamp}`);

  const workspace =
    process.env.SPORE_E2E_WORKSPACE ??
    join(tmpdir(), `spore-e2e-${sessionId.asString()}`);
  mkdirSync(workspace, { recursive: true });
  prepareWorkspace(scenario, workspace);
  process.stdout.write(`workspace  : ${workspace}\n`);

  if (!mock && !process.env.SPORE_OTLP_ENDPOINT) {
    process.stderr.write(
      "note: SPORE_OTLP_ENDPOINT is unset — writing JSONL only (no Tempo forwarding).\n",
    );
  }

  const result = mock
    ? await runMock(scenario, sessionId)
    : await runLive(scenario, sessionId, modelId, workspace);

  if (result.kind === "success") {
    process.stdout.write(`result     : Success (${result.turns} turns)\n`);
    process.stdout.write(`output     : ${JSON.stringify(result.output)}\n`);
  } else {
    process.stderr.write(`result     : ${JSON.stringify(result)}\n`);
    process.exit(1);
  }
}

async function runLive(
  scenario: ScenarioId,
  sessionId: SessionId,
  modelId: string,
  workspace: string,
): Promise<RunResult> {
  const baseUrl = process.env.SPORE_OLLAMA_BASE_URL ?? OLLAMA_DEFAULT_BASE_URL;
  const model = OllamaModelInterface.withBaseUrl(modelId, baseUrl);
  const agent: Agent = new ModelAgent(AgentId.of("e2e-agent"), model);

  const registry = buildRealToolRegistry(scenario);
  const sandbox: SandboxProvider = new WorkspaceScopedSandbox({
    root: workspace,
    allowed_paths: [],
    denied_paths: [],
    denied_extensions: [],
    read_only: false,
    max_file_size: 0,
  });
  const bridge = new RealToolRegistry(
    registry,
    sandbox,
    sessionId,
    new coreStorage.InMemoryStorageProvider(),
    new coreStorage.InMemoryStorageProvider(),
  );
  const toolSchemas = bridge.modelSchemas();
  const tools: HarnessToolRegistry = bridge;

  const windowLimit = scenario === "s3" ? 200 : 128_000;
  const cfg: coreContext.CompactionConfig = {
    threshold: 0.8,
    preserve_recent_n: 2,
    head_tail_tokens: 64,
    offload_path: join(workspace, ".spore/offload"),
    max_compaction_attempts: 2,
  };
  const contextManager: HarnessContextManager = buildRichContextManager(
    model,
    new OllamaCacheProvider(),
    cfg,
  );

  const termination: TerminationPolicy = new CompleteOnFinalResponse();
  const obs = new OutboxObservabilityProvider(
    outboxConfig(join(workspace, ".spore"), { flushOnSessionEnd: true }),
  );
  process.stdout.write(`trace_id   : ${obs.traceIdFor(sessionId)}\n`);

  const harness = buildScenario({
    scenario,
    agent,
    tools,
    sandbox,
    contextManager,
    terminationPolicy: termination,
    toolSchemas,
    observability: obs,
  });

  return runScenario(scenario, harness, sessionId, windowLimit);
}

async function runScenario(
  scenario: ScenarioId,
  harness: { run(o: HarnessRunOptions): Promise<RunResult> },
  sessionId: SessionId,
  windowLimit: number,
): Promise<RunResult> {
  if (scenario === "s2") {
    const task1 = newTask(scenarioPrompt("s2"), sessionId, reactStrategy(8));
    const r1 = await harness.run({ task: task1 });
    if (r1.kind !== "success") {
      process.stderr.write(
        `S2 turn 1 did not succeed: ${JSON.stringify(r1)}\n`,
      );
      return r1;
    }
    const task2 = newTask(
      "Add a second TODO item to notes.md that references the first item you wrote. " +
        "Use write_file with append=true. Reply DONE when finished.",
      sessionId,
      reactStrategy(8),
    );
    return harness.run({ task: task2, session_state: emptySessionState() });
  }

  if (scenario === "s3") {
    const task = newTask(scenarioPrompt("s3"), sessionId, reactStrategy(8));
    const state: SessionState = emptySessionState();
    seedCompactionState(
      state,
      "deploy the payment service",
      sessionId,
      task.id as TaskId,
      windowLimit,
      Math.floor(windowLimit * 0.82),
      12,
    );
    return harness.run({ task, session_state: state });
  }

  const task = newTask(scenarioPrompt(scenario), sessionId, reactStrategy(8));
  return harness.run({ task });
}

function prepareWorkspace(scenario: ScenarioId, workspace: string): void {
  if (scenario === "s1" || scenario === "s5") {
    writeFileSync(
      join(workspace, "input.txt"),
      "hello from the spore harness end to end scenario\n",
    );
  }
}

// ---------------------------------------------------------------------------
// Offline / mock mode.
// ---------------------------------------------------------------------------

async function runMock(
  scenario: ScenarioId,
  sessionId: SessionId,
): Promise<RunResult> {
  const { MockAgent } = await import("@spore/core");
  const { harnessTesting } = await import("@spore/core");
  const agent = new MockAgent(AgentId.of("mock"));
  const usage = { input_tokens: 10, output_tokens: 5 };
  agent.push({
    kind: "tool_call_requested",
    calls: [{ id: "c1", name: "read_file", input: { path: "input.txt" } }],
    usage,
  });
  agent.push({ kind: "final_response", content: "DONE", usage });

  const tools = new harnessTesting.ScriptedToolRegistry();
  tools.push({ kind: "success", content: "contents", truncated: false });

  const harness = buildScenario({
    scenario,
    agent,
    tools,
    sandbox: new harnessTesting.AllowAllSandbox(),
    contextManager: new harnessTesting.NoopContextManager(),
    terminationPolicy: new harnessTesting.AlwaysContinuePolicy(),
    toolSchemas: [] as ToolSchema[],
  });
  const task = newTask(scenarioPrompt(scenario), sessionId, reactStrategy(5));
  return harness.run({ task });
}

void main();
